### ISSUE TITLE
glfwPlatformPollEvents can hang, assumes other code won't pull XEvents

### ISSUE THREAD

**Issue by johnjosephorr** [glfw/glfw#1225]
from master/src/x11_window.c:

```c++
void _glfwPlatformPollEvents(void)
{
    _GLFWwindow* window;

#if defined(__linux__)
    _glfwDetectJoystickConnectionLinux();
#endif
    int count = XPending(_glfw.x11.display);
    while (count--)
    {
        XEvent event;
        XNextEvent(_glfw.x11.display, &event);
        processEvent(&event);
    }
```
This code assumes that it will be the only code removing the "count" number of XEvent's from the queue (it assumes it can call XNextEvent "count" number of times and it will not block waiting for input).  If something else should pull an XEvent off the queue (or perhaps Xlib will cancel events or coalesce multiple events that are currently on the queue?) this code will call XNextEvent more times than it should to process the events that are currently on the queue and will block inside XNextEvent until something generates new events.

On a normal computer there is a person with a mouse and keyboard who will typically generate more events, but on a system driving a non-interactive display of some kind, there is no one at the keyboard or mouse, so this code can hang.

In my case, I modified this loop to check XPending before calling XNextEvent:

```c++
    while (count--)
    {
        if ( XPending(_glfw.x11.display) > 0 )
        {
            XEvent event;
            XNextEvent(_glfw.x11.display, &event);
            processEvent(&event);
        }
    }
```

I intentionally left the cap in here on "count" number of loops so it does not process more than the original set of events, which matches the original behavior.  

If it matters, I ran into this while using Cinder which uses glfw in Linux only.

**Comment by ILGSupreme** [comment]
Hello, it seems i have encountered this problem when building my application. Whenever i tried to exit the application i get this error. For now i have only encountered this on Ubuntu 16.04 LT and not on Windows.

I followed the steps provided by GLFW introduction so i dont think it is on my end , might be...

The error comes after a callback from _glfwInputKey. The callback itselfs just calls glfwDestroyWindow(window), and glfwTerminate in order. then exits the call..

**Comment by tombsar** [comment]
@ILGSupreme Sorry, calling either `glfwDestroyWindow` or `glfwTerminate` from within a callback is explicitly disallowed in the current API. Please see http://www.glfw.org/docs/latest/intro_guide.html#reentrancy

**Comment by ILGSupreme** [comment]
thank you very much! my issue is now resolved. Have a happy weekend!

**Comment by elmindreda** [comment]
@johnjosephorr To clarify, this is an issue you have encountered and not just a theoretical concern?

**Comment by johnjosephorr** [comment]
I ran into this on a system that was intended to run without any local input devices (no mouse or keyboard).

**Comment by elmindreda** [comment]
@johnjosephorr Thank you. Will fix.

**Comment by elmindreda** [comment]
This should hopefully be fixed now with 8d1a64c83180fd8c5e41b5f2af3416d604261383.
